patches/server/0579-Add-StructureLocateEvent.patch

```diff
@@ -5,27 +5,138 @@ Subject: [PATCH] Add StructureLocateEvent
+index f9fc2fc63080a60fe61ebb08ddd93c4f189df84d..b8770a888209353421cdcc3dbac610e415b72fa1 100644
 --- a/src/main/java/net/minecraft/world/level/chunk/ChunkGenerator.java
 +++ b/src/main/java/net/minecraft/world/level/chunk/ChunkGenerator.java
-@@ -294,6 +294,20 @@ public abstract class ChunkGenerator implements BiomeManager.NoiseBiomeSource {
+@@ -294,6 +294,33 @@ public abstract class ChunkGenerator implements BiomeManager.NoiseBiomeSource {
  
      @Nullable
      public Pair<BlockPos, Holder<ConfiguredStructureFeature<?, ?>>> findNearestMapFeature(ServerLevel worldserver, HolderSet<ConfiguredStructureFeature<?, ?>> holderset, BlockPos center, int radius, boolean skipExistingChunks) {
-+        // Paper start
-+        /*org.bukkit.World world1 = worldserver.getWorld();
-+        org.bukkit.Location originLocation = new org.bukkit.Location(world1, center.getX(), center.getY(), center.getZ());
-+        io.papermc.paper.event.world.StructureLocateEvent event = new io.papermc.paper.event.world.StructureLocateEvent(world1, originLocation, org.bukkit.StructureType.getStructureTypes().get(structureFeature.getFeatureName()), radius, skipExistingChunks);
-+        if(!event.callEvent()) return null;
-+        // If event call set a final location, skip structure finding and just return set result.
-+        if(event.getResult() != null) return new BlockPos(event.getResult().getBlockX(), event.getResult().getBlockY(), event.getResult().getBlockZ());
-+        // Get origin location (re)defined by event call.
-+        center = new BlockPos(event.getOrigin().getBlockX(), event.getOrigin().getBlockY(), event.getOrigin().getBlockZ());
-+        // Get radius and whether to find unexplored structures (re)defined by event call.
++        // Paper start - StructureLocateEvent
++        final org.bukkit.World world = worldserver.getWorld();
++        final org.bukkit.Location origin = net.minecraft.server.MCUtil.toLocation(worldserver, center);
++        final Registry<ConfiguredStructureFeature<?, ?>> configuredStructureFeatureRegistry = worldserver.registryAccess().registryOrThrow(Registry.CONFIGURED_STRUCTURE_FEATURE_REGISTRY);
++        final List<io.papermc.paper.world.structure.ConfiguredStructure> configuredStructures = new ArrayList<>();
++        for (Holder<net.minecraft.world.level.levelgen.feature.ConfiguredStructureFeature<?, ?>> holder : holderset) {
++            final net.minecraft.resources.ResourceLocation configuredStructureResourceLoc = configuredStructureFeatureRegistry.getKey(holder.value());
++            if (configuredStructureResourceLoc == null) {
++                continue;
++            }
++            final io.papermc.paper.world.structure.ConfiguredStructure configuredStructure = org.bukkit.Registry.CONFIGURED_STRUCTURE.get(org.bukkit.craftbukkit.util.CraftNamespacedKey.fromMinecraft(configuredStructureResourceLoc));
++            if (configuredStructure != null) {
```

Comment:
should this maybe warn if it's not found? seems bad if that info/context was lost

Reply:
No, because datapacks can add their own configured structures and those wouldn't have API representations. So I think this is a perfectly plausible scenario.

Reply:
This can be mitigated if ConfiguredStructure moved to a PaperConfiguredStructure -> ConfiguredStructure w/ConfiguredStructureWrapper system like enchantments.

Then a new instance of PaperConfiguredStructure could be ctor'd when a datapack creates one.